scylla-cql/src/_macro_internal.rs

```diff
@@ -34,17 +34,57 @@ pub use crate::serialize::value::{
 pub use crate::serialize::writers::WrittenCellProof;
 pub use crate::serialize::{CellValueBuilder, CellWriter, RowWriter, SerializationError};
 
+/// Represents a set of values that can be sent along a CQL statement when serializing by name
+///
+/// For now this trait is an implementation detail of `#[derive(SerializeRow)]` when
+/// serializing by name
+pub trait SerializeRowByName {
+    /// A type that can handle serialization of this struct column-by-column
+    type Partial<'d>: PartialSerializeRowByName
+    where
+        Self: 'd;
+
+    /// Returns a type that can serialize this row "column-by-column"
+    fn partial(&self) -> Self::Partial<'_>;
+}
+
+/// How to serialize a row column-by-column
+///
+/// For now this trait is an implementation detail of `#[derive(SerializeRow)]` when
+/// serializing by name
+pub trait PartialSerializeRowByName {
+    /// Tries to serialize a single column in the row according to the information in the given
+    /// context.
+    ///
+    /// It returns whether the column finished the serialization of the struct, did it partially,
+    /// was not used at all, or errored.
+    fn serialize_field(
+        &mut self,
+        spec: &ColumnSpec,
+        writer: &mut RowWriter<'_>,
+    ) -> Result<self::ser::row::FieldStatus, SerializationError>;
+
+    /// Checks if there are any missing columns to finish the serialization
+    fn check_missing(self) -> Result<(), SerializationError>;
+}
```

Comment:
❓ I'd just like to check if I understand this well:
check_missing will return `Ok` if either:
- some previous call to `serialize_field` returned `Ok(Done)`
- struct is empty

Otherwise it will return an error.

If that is not correct, please describe relation between the methods a bit more so that the reader can build some intuition.
If it is correct, the reasons for `check_missing` are:
- Empty structs
- Generating error messages
because those cases are the only things not covered by remembering if `Ok(Done)` wes returned before.
Am I right, or do I misunderstand the purpose of this?

Reply:
Ok, thanks!